File: src/platform/macos/glyph_texture_smoke.m
#import <CoreGraphics/CoreGraphics.h>
#import <CoreText/CoreText.h>
#import <Foundation/Foundation.h>
#import <Metal/Metal.h>
#include <stdint.h>
#include <stddef.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    int32_t status;
    uint32_t metal_device_created;
    uint32_t command_queue_created;
    uint32_t source_rasterized;
    uint32_t texture_created;
    uint32_t texture_uploaded;
    uint32_t texture_readback;
    uint32_t source_width;
    uint32_t source_height;
    uint32_t source_non_clear_pixels;
    uint32_t texture_non_clear_pixels;
    uint32_t texture_mismatched_pixels;
    uint32_t upload_bytes;
    uint32_t readback_failures;
} MaruGlyphTextureSmokeResult;

typedef struct {
    uint8_t *pixels;
    size_t width;
    size_t height;
    size_t bytes_per_row;
    size_t byte_count;
    uint32_t non_clear_pixels;
} MaruGlyphTextureBitmap;

static void maru_clear_result(MaruGlyphTextureSmokeResult *result) {
    result->status = -1;
    result->metal_device_created = 0;
    result->command_queue_created = 0;
    result->source_rasterized = 0;
    result->texture_created = 0;
    result->texture_uploaded = 0;
    result->texture_readback = 0;
    result->source_width = 0;
    result->source_height = 0;
    result->source_non_clear_pixels = 0;
    result->texture_non_clear_pixels = 0;
    result->texture_mismatched_pixels = 0;
    result->upload_bytes = 0;
    result->readback_failures = 0;
}

static CFStringRef maru_create_probe_string(void) {
    static const UInt8 bytes[] = { 'M', 'a', 'r', 'u' };
    return CFStringCreateWithBytes(
        kCFAllocatorDefault,
        bytes,
        sizeof(bytes),
        kCFStringEncodingUTF8,
        false
    );
}

static CTFontRef maru_create_primary_font(void) {
    CTFontRef font = CTFontCreateUIFontForLanguage(kCTFontUIFontUserFixedPitch, 24.0, NULL);
    if (font != NULL) {
        return font;
    }
    return CTFontCreateWithName(CFSTR("Menlo-Regular"), 24.0, NULL);
}

static void maru_free_bitmap(MaruGlyphTextureBitmap *bitmap) {
    if (bitmap->pixels != NULL) {
        free(bitmap->pixels);
    }
    bitmap->pixels = NULL;
    bitmap->width = 0;
    bitmap->height = 0;
    bitmap->bytes_per_row = 0;
    bitmap->byte_count = 0;
    bitmap->non_clear_pixels = 0;
}

static uint32_t maru_count_non_clear_rgba_pixels(const uint8_t *pixels, size_t byte_count) {
    uint32_t non_clear = 0;
    for (size_t offset = 0; offset + 3 < byte_count; offset += 4) {
        if (pixels[offset + 3] != 0) {
            non_clear += 1;
        }
    }
    return non_clear;
}

static uint32_t maru_count_mismatched_bytes(
    const uint8_t *left,
    const uint8_t *right,
    size_t byte_count
) {
    uint32_t mismatches = 0;
    for (size_t index = 0; index < byte_count; index++) {
        if (left[index] != right[index]) {
            mismatches += 1;
        }
    }
    return mismatches;
}

static int maru_make_coretext_bitmap(
    MaruGlyphTextureSmokeResult *result,
    MaruGlyphTextureBitmap *bitmap
) {
    // 이 bitmap은 제품 atlas가 아니다. CoreText가 만든 실제 glyph pixels를 Metal
    // texture upload smoke의 입력으로 쓰기 위한 작은 fixture다. 제품 renderer가
    // 들어오면 atlas packing/eviction은 별도 모듈에서 소유한다.
    const size_t width = 128;
    const size_t height = 64;
    const size_t bytes_per_pixel = 4;
    const size_t bytes_per_row = width * bytes_per_pixel;
    const size_t byte_count = bytes_per_row * height;

    bitmap->pixels = (uint8_t *)calloc(byte_count, 1);
    if (bitmap->pixels == NULL) {
        return 2;
    }
    bitmap->width = width;
    bitmap->height = height;
    bitmap->bytes_per_row = bytes_per_row;
    bitmap->byte_count = byte_count;

    CTFontRef font = maru_create_primary_font();
    if (font == NULL) {
        return 2;
    }

    CFStringRef probe = maru_create_probe_string();
    if (probe == NULL) {
        CFRelease(font);
        return 2;
    }

    const void *keys[] = { kCTFontAttributeName };
    const void *values[] = { font };
    CFDictionaryRef attributes = CFDictionaryCreate(
        kCFAllocatorDefault,
        keys,
        values,
        1,
        &kCFTypeDictionaryKeyCallBacks,
        &kCFTypeDictionaryValueCallBacks
    );
    if (attributes == NULL) {
        CFRelease(probe);
        CFRelease(font);
        return 2;
    }

    CFAttributedStringRef attributed = CFAttributedStringCreate(
        kCFAllocatorDefault,
        probe,
        attributes
    );
    if (attributed == NULL) {
        CFRelease(attributes);
        CFRelease(probe);
        CFRelease(font);
        return 2;
    }

    CTLineRef line = CTLineCreateWithAttributedString(attributed);
    if (line == NULL) {
        CFRelease(attributed);
        CFRelease(attributes);
        CFRelease(probe);
        CFRelease(font);
        return 2;
    }

    CGColorSpaceRef color_space = CGColorSpaceCreateDeviceRGB();
    if (color_space == NULL) {
        CFRelease(line);
        CFRelease(attributed);
        CFRelease(attributes);
        CFRelease(probe);
        CFRelease(font);
        return 2;
    }

    CGContextRef context = CGBitmapContextCreate(
        bitmap->pixels,
        width,
        height,
        8,
        bytes_per_row,
        color_space,
        kCGImageAlphaPremultipliedLast | kCGBitmapByteOrder32Big
    );
    if (context == NULL) {
        CGColorSpaceRelease(color_space);
        CFRelease(line);
        CFRelease(attributed);
        CFRelease(attributes);
        CFRelease(probe);
        CFRelease(font);
        return 2;
    }

    CGContextSetTextMatrix(context, CGAffineTransformIdentity);
    CGContextTranslateCTM(context, 0.0, (CGFloat)height);
    CGContextScaleCTM(context, 1.0, -1.0);
    CGContextSetShouldAntialias(context, true);
    CGContextSetRGBFillColor(context, 1.0, 1.0, 1.0, 1.0);
    CGContextSetTextPosition(context, 8.0, 36.0);
    CTLineDraw(line, context);

    bitmap->non_clear_pixels = maru_count_non_clear_rgba_pixels(bitmap->pixels, bitmap->byte_count);
    result->source_width = (uint32_t)width;
    result->source_height = (uint32_t)height;
    result->source_non_clear_pixels = bitmap->non_clear_pixels;
    result->source_rasterized = bitmap->non_clear_pixels > 0 ? 1 : 0;

    CGContextRelease(context);
    CGColorSpaceRelease(color_space);
    CFRelease(line);
    CFRelease(attributed);
    CFRelease(attributes);
    CFRelease(probe);
    CFRelease(font);

    return bitmap->non_clear_pixels > 0 ? 0 : 2;
}

static int maru_upload_and_readback_texture(
    MaruGlyphTextureSmokeResult *result,
    const MaruGlyphTextureBitmap *bitmap
) {
    id<MTLDevice> device = MTLCreateSystemDefaultDevice();
    if (device == nil) {
        return 3;
    }
    result->metal_device_created = 1;

    id<MTLCommandQueue> queue = [device newCommandQueue];
    if (queue == nil) {
        return 4;
    }
    result->command_queue_created = 1;

    MTLTextureDescriptor *descriptor = [MTLTextureDescriptor
        texture2DDescriptorWithPixelFormat:MTLPixelFormatRGBA8Unorm
                                     width:bitmap->width
                                    height:bitmap->height
                                 mipmapped:NO];
    descriptor.storageMode = MTLStorageModeShared;
    descriptor.usage = MTLTextureUsageShaderRead;

    id<MTLTexture> texture = [device newTextureWithDescriptor:descriptor];
    if (texture == nil) {
        return 5;
    }
    result->texture_created = 1;

    [texture replaceRegion:MTLRegionMake2D(0, 0, bitmap->width, bitmap->height)
               mipmapLevel:0
                 withBytes:bitmap->pixels
               bytesPerRow:bitmap->bytes_per_row];
    result->texture_uploaded = 1;
    // upload_bytes는 "실제로 texture에 올린 byte 수"를 뜻해야 한다. 그래서 source bitmap을
    // 만들 때가 아니라 replaceRegion이 끝난 직후에 기록한다. device가 없어 upload가
    // 일어나지 않은 run에서는 0으로 남아, summary만 보고 "GPU에 올린 게 없다"를 구분할 수 있다.
    result->upload_bytes = (bitmap->byte_count > UINT32_MAX) ? UINT32_MAX : (uint32_t)bitmap->byte_count;

    id<MTLBuffer> readback_buffer = [device
        newBufferWithLength:bitmap->byte_count
                    options:MTLResourceStorageModeShared];
    if (readback_buffer == nil) {
        result->readback_failures += 1;
        return 6;
    }

    id<MTLCommandBuffer> command_buffer = [queue commandBuffer];
    if (command_buffer == nil) {
        result->readback_failures += 1;
        return 6;
    }

    id<MTLBlitCommandEncoder> blit = [command_buffer blitCommandEncoder];
    if (blit == nil) {
        result->readback_failures += 1;
        return 6;
    }

    [blit copyFromTexture:texture
              sourceSlice:0
              sourceLevel:0
             sourceOrigin:MTLOriginMake(0, 0, 0)
               sourceSize:MTLSizeMake(bitmap->width, bitmap->height, 1)
                 toBuffer:readback_buffer
        destinationOffset:0
   destinationBytesPerRow:bitmap->bytes_per_row
 destinationBytesPerImage:bitmap->byte_count];
    [blit endEncoding];

    [command_buffer commit];
    [command_buffer waitUntilCompleted];

    const uint8_t *readback = (const uint8_t *)[readback_buffer contents];
    if (readback == NULL) {
        result->readback_failures += 1;
        return 6;
    }

    result->texture_readback = 1;
    result->texture_non_clear_pixels = maru_count_non_clear_rgba_pixels(readback, bitmap->byte_count);
    result->texture_mismatched_pixels = maru_count_mismatched_bytes(
        bitmap->pixels,
        readback,
        bitmap->byte_count
    );

    return result->texture_non_clear_pixels == result->source_non_clear_pixels &&
            result->texture_mismatched_pixels == 0
        ? 0
        : 7;
}

// result->status는 어느 단계에서 멈췄는지 summary(native_status/native_status_label)로
// 분리하기 위한 값이다. 새 값을 추가하면 glyph_texture_smoke.zig의 nativeStatusLabel도
// 같이 갱신해야 artifact와 native bridge가 같은 말을 한다:
//   0  = 성공(raster + upload + readback + byte 일치)
//   2  = CoreText/CoreGraphics CPU bitmap raster 실패(폰트/컨텍스트 생성 또는 0-ink)
//   3  = Metal device 생성 실패(headless/GPU 없음)
//   4  = command queue 생성 실패
//   5  = texture 생성 실패(예: storage mode 미지원)
//   6  = readback 인프라 실패(buffer/command buffer/blit encoder/contents가 nil)
//   7  = readback 픽셀이 source bitmap과 불일치
//  -1  = 아직 실행되지 않음
void maru_macos_glyph_texture_smoke_run(MaruGlyphTextureSmokeResult *result) {
    @autoreleasepool {
        if (result == NULL) {
            return;
        }
        maru_clear_result(result);

        MaruGlyphTextureBitmap bitmap = {0};
        int status = maru_make_coretext_bitmap(result, &bitmap);
        if (status == 0) {
            status = maru_upload_and_readback_texture(result, &bitmap);
        }

        result->status = status;
        maru_free_bitmap(&bitmap);
    }
}

File: src/platform/macos/glyph_texture_smoke.zig
const std = @import("std");

const artifact_dir = "zig-out/maru-macos-glyph-texture-smoke";

const NativeGlyphTextureSmokeResult = extern struct {
    status: c_int,
    metal_device_created: u32,
    command_queue_created: u32,
    source_rasterized: u32,
    texture_created: u32,
    texture_uploaded: u32,
    texture_readback: u32,
    source_width: u32,
    source_height: u32,
    source_non_clear_pixels: u32,
    texture_non_clear_pixels: u32,
    texture_mismatched_pixels: u32,
    upload_bytes: u32,
    readback_failures: u32,
};

extern fn maru_macos_glyph_texture_smoke_run(result: *NativeGlyphTextureSmokeResult) void;

pub fn main(init: std.process.Init) !void {
    const io = init.io;
    const allocator = init.gpa;

    var stdout_buffer: [1024]u8 = undefined;
    var stdout_file_writer: std.Io.File.Writer = .init(.stdout(), io, &stdout_buffer);
    const stdout = &stdout_file_writer.interface;

    var native: NativeGlyphTextureSmokeResult = emptyNativeResult();
    maru_macos_glyph_texture_smoke_run(&native);

    const smoke_status = deriveSmokeStatus(native);
    const summary = try renderSummary(allocator, smoke_status, native);
    defer allocator.free(summary);

    try writeSummary(io, summary);
    try stdout.writeAll(summary);
    try stdout.print("\nartifacts written to {s}/\n", .{artifact_dir});
    try stdout.flush();

    if (!smoke_status.glyph_texture_uploaded) return error.MacosGlyphTextureSmokeFailed;
}

fn emptyNativeResult() NativeGlyphTextureSmokeResult {
    return .{
        .status = -1,
        .metal_device_created = 0,
        .command_queue_created = 0,
        .source_rasterized = 0,
        .texture_created = 0,
        .texture_uploaded = 0,
        .texture_readback = 0,
        .source_width = 0,
        .source_height = 0,
        .source_non_clear_pixels = 0,
        .texture_non_clear_pixels = 0,
        .texture_mismatched_pixels = 0,
        .upload_bytes = 0,
        .readback_failures = 0,
    };
}

const SmokeStatus = struct {
    source_rasterized: bool,
    metal_texture: bool,
    glyph_texture_uploaded: bool,
};

fn deriveSmokeStatus(native: NativeGlyphTextureSmokeResult) SmokeStatus {
    // 이 smoke는 글자를 화면에 그리지 않는다. CoreText가 만든 CPU bitmap이 Metal
    // texture로 올라가고, 같은 texture를 다시 읽었을 때 픽셀이 보존되는지만 본다.
    // 이렇게 해야 다음 text draw 실패가 rasterizer 문제인지 texture upload 문제인지
    // 작은 summary만 보고 분리할 수 있다.
    const source_rasterized = native.source_rasterized != 0 and
        native.source_width > 0 and
        native.source_height > 0 and
        native.source_non_clear_pixels > 0;
    const metal_texture = native.metal_device_created != 0 and
        native.command_queue_created != 0 and
        native.texture_created != 0;
    const glyph_texture_uploaded = source_rasterized and
        metal_texture and
        native.status == 0 and
        native.texture_uploaded != 0 and
        native.texture_readback != 0 and
        native.texture_non_clear_pixels == native.source_non_clear_pixels and
        native.texture_mismatched_pixels == 0 and
        native.upload_bytes > 0 and
        native.readback_failures == 0;

    return .{
        .source_rasterized = source_rasterized,
        .metal_texture = metal_texture,
        .glyph_texture_uploaded = glyph_texture_uploaded,
    };
}

fn renderSummary(
    allocator: std.mem.Allocator,
    smoke_status: SmokeStatus,
    native: NativeGlyphTextureSmokeResult,
) ![]u8 {
    var output: std.Io.Writer.Allocating = .init(allocator);
    errdefer output.deinit();

    const writer = &output.writer;
    try writer.writeAll("maru.macos-glyph-texture-smoke.v1\n");
    try writer.print("artifact_dir={s}\n", .{artifact_dir});
    try writer.print("source_rasterized={}\n", .{smoke_status.source_rasterized});
    try writer.print("metal_texture={}\n", .{smoke_status.metal_texture});
    try writer.print("glyph_texture_uploaded={}\n", .{smoke_status.glyph_texture_uploaded});
    try writer.writeAll("glyph_text=false\n");
    try writer.writeAll("ui_note=coretext_cpu_bitmap_to_metal_texture_no_window_no_text_draw\n");
    // artifact만 보고도 실패 단계를 읽을 수 있게 숫자 status와 사람이 읽는 label을
    // 같이 남긴다. 새 native status를 추가할 때는 bridge의 status 범례와 이 label을
    // 함께 갱신해야 한다.
    try writer.print("native_status={d}\n", .{native.status});
    try writer.print("native_status_label={s}\n", .{nativeStatusLabel(native.status)});
    try writer.print("metal_device_created={d}\n", .{native.metal_device_created});
    try writer.print("command_queue_created={d}\n", .{native.command_queue_created});
    try writer.print("native_source_rasterized={d}\n", .{native.source_rasterized});
    try writer.print("texture_created={d}\n", .{native.texture_created});
    try writer.print("texture_uploaded={d}\n", .{native.texture_uploaded});
    try writer.print("texture_readback={d}\n", .{native.texture_readback});
    try writer.print("source_width={d}\n", .{native.source_width});
    try writer.print("source_height={d}\n", .{native.source_height});
    try writer.print("source_non_clear_pixels={d}\n", .{native.source_non_clear_pixels});
    try writer.print("texture_non_clear_pixels={d}\n", .{native.texture_non_clear_pixels});
    try writer.print("texture_mismatched_pixels={d}\n", .{native.texture_mismatched_pixels});
    try writer.print("upload_bytes={d}\n", .{native.upload_bytes});
    try writer.print("readback_failures={d}\n", .{native.readback_failures});

    return output.toOwnedSlice();
}

fn nativeStatusLabel(status: c_int) []const u8 {
    return switch (status) {
        -1 => "not_run",
        0 => "ok",
        2 => "cpu_raster_failed",
        3 => "metal_device_failed",
        4 => "command_queue_failed",
        5 => "texture_create_failed",
        6 => "readback_infra_failed",
        7 => "readback_mismatch",
        else => "unknown",
    };
}

fn writeSummary(io: std.Io, summary: []const u8) !void {
    try std.Io.Dir.cwd().createDirPath(io, artifact_dir);
    try std.Io.Dir.cwd().writeFile(io, .{
        .sub_path = artifact_dir ++ "/glyph-texture.summary.txt",
        .data = summary,
        .flags = .{ .truncate = true },
    });
}

test "macOS glyph texture smoke summary reports upload boundary" {
    // native Metal을 호출하지 않는 계약 테스트에서도 summary 의미를 고정한다.
    // 그래야 texture upload 구현이 바뀌어도 artifact consumer가 무엇을 읽어야 하는지
    // 흔들리지 않는다.
    const native: NativeGlyphTextureSmokeResult = .{
        .status = 0,
        .metal_device_created = 1,
        .command_queue_created = 1,
        .source_rasterized = 1,
        .texture_created = 1,
        .texture_uploaded = 1,
        .texture_readback = 1,
        .source_width = 128,
        .source_height = 64,
        .source_non_clear_pixels = 77,
        .texture_non_clear_pixels = 77,
        .texture_mismatched_pixels = 0,
        .upload_bytes = 32768,
        .readback_failures = 0,
    };
    const summary = try renderSummary(std.testing.allocator, deriveSmokeStatus(native), native);
    defer std.testing.allocator.free(summary);

    try std.testing.expect(std.mem.indexOf(u8, summary, "maru.macos-glyph-texture-smoke.v1\n") != null);
    try std.testing.expect(std.mem.indexOf(u8, summary, "source_rasterized=true\n") != null);
    try std.testing.expect(std.mem.indexOf(u8, summary, "metal_texture=true\n") != null);
    try std.testing.expect(std.mem.indexOf(u8, summary, "glyph_texture_uploaded=true\n") != null);
    try std.testing.expect(std.mem.indexOf(u8, summary, "glyph_text=false\n") != null);
    try std.testing.expect(std.mem.indexOf(u8, summary, "native_status=0\n") != null);
    try std.testing.expect(std.mem.indexOf(u8, summary, "native_status_label=ok\n") != null);
    try std.testing.expect(std.mem.indexOf(u8, summary, "source_non_clear_pixels=77\n") != null);
    try std.testing.expect(std.mem.indexOf(u8, summary, "texture_non_clear_pixels=77\n") != null);
    try std.testing.expect(std.mem.indexOf(u8, summary, "texture_mismatched_pixels=0\n") != null);
    try std.testing.expect(std.mem.indexOf(u8, summary, "upload_bytes=32768\n") != null);
}

test "glyph texture smoke status labels explain native failure stages" {
    // 실패 artifact를 볼 때 숫자 status만 있으면 bridge 코드를 다시 열어봐야 한다.
    // label 계약을 고정해 CI artifact만으로도 어느 단계에서 멈췄는지 바로 읽게 한다.
    try std.testing.expectEqualStrings("not_run", nativeStatusLabel(-1));
    try std.testing.expectEqualStrings("ok", nativeStatusLabel(0));
    try std.testing.expectEqualStrings("cpu_raster_failed", nativeStatusLabel(2));
    try std.testing.expectEqualStrings("metal_device_failed", nativeStatusLabel(3));
    try std.testing.expectEqualStrings("command_queue_failed", nativeStatusLabel(4));
    try std.testing.expectEqualStrings("texture_create_failed", nativeStatusLabel(5));
    try std.testing.expectEqualStrings("readback_infra_failed", nativeStatusLabel(6));
    try std.testing.expectEqualStrings("readback_mismatch", nativeStatusLabel(7));
    try std.testing.expectEqualStrings("unknown", nativeStatusLabel(99));
}

test "glyph texture upload requires readback to match the source bitmap" {
    // texture를 만들었다는 사실만으로는 glyph pixel이 제대로 올라갔다고 말할 수 없다.
    // readback 결과가 source bitmap의 non-clear 픽셀 수와 byte 내용까지 맞아야 한다.
    var partial = emptyNativeResult();
    partial.status = 0;
    partial.metal_device_created = 1;
    partial.command_queue_created = 1;
    partial.source_rasterized = 1;
    partial.texture_created = 1;
    partial.texture_uploaded = 1;
    partial.texture_readback = 1;
    partial.source_width = 128;
    partial.source_height = 64;
    partial.source_non_clear_pixels = 77;
    partial.texture_non_clear_pixels = 76;
    partial.texture_mismatched_pixels = 0;
    partial.upload_bytes = 32768;
    try std.testing.expect(!deriveSmokeStatus(partial).glyph_texture_uploaded);

    var mismatch = partial;
    mismatch.texture_non_clear_pixels = 77;
    mismatch.texture_mismatched_pixels = 1;
    try std.testing.expect(!deriveSmokeStatus(mismatch).glyph_texture_uploaded);

    var success = mismatch;
    success.texture_mismatched_pixels = 0;
    try std.testing.expect(deriveSmokeStatus(success).glyph_texture_uploaded);
}

test "glyph texture smoke fails closed on each raster, GPU, and readback stage" {
    // 이 smoke의 목적은 "CPU raster는 됐는데 GPU upload/readback에서 막혔다"를 summary만
    // 보고 분리하는 것이다. 그래서 성공 모양에서 단계별 신호를 하나씩 뒤집어, gate의 각
    // 항이 실제로 부하를 받는지(없으면 false로 닫히는지) 고정한다. native_status 코드
    // 의미는 glyph_texture_smoke.m의 범례를 참조한다.
    const ok: NativeGlyphTextureSmokeResult = .{
        .status = 0,
        .metal_device_created = 1,
        .command_queue_created = 1,
        .source_rasterized = 1,
        .texture_created = 1,
        .texture_uploaded = 1,
        .texture_readback = 1,
        .source_width = 128,
        .source_height = 64,
        .source_non_clear_pixels = 77,
        .texture_non_clear_pixels = 77,
        .texture_mismatched_pixels = 0,
        .upload_bytes = 32768,
        .readback_failures = 0,
    };
    // 기준선: 성공 모양은 통과해야 한다.
    try std.testing.expect(deriveSmokeStatus(ok).glyph_texture_uploaded);

    // GPU가 없는 headless 실행(device 생성 실패, status 3): metal_texture=false이고
    // 전체 gate도 false여서 프로세스가 non-zero로 종료된다.
    var no_device = ok;
    no_device.metal_device_created = 0;
    no_device.status = 3;
    try std.testing.expect(!deriveSmokeStatus(no_device).metal_texture);
    try std.testing.expect(!deriveSmokeStatus(no_device).glyph_texture_uploaded);

    // command queue 생성 실패(status 4).
    var no_queue = ok;
    no_queue.command_queue_created = 0;
    no_queue.status = 4;
    try std.testing.expect(!deriveSmokeStatus(no_queue).metal_texture);
    try std.testing.expect(!deriveSmokeStatus(no_queue).glyph_texture_uploaded);

    // texture 생성 실패(status 5, 예: storage mode 미지원).
    var no_texture = ok;
    no_texture.texture_created = 0;
    no_texture.status = 5;
    try std.testing.expect(!deriveSmokeStatus(no_texture).metal_texture);
    try std.testing.expect(!deriveSmokeStatus(no_texture).glyph_texture_uploaded);

    // 단계 status가 비-0이면(예: 내부 단계 실패) 다른 신호가 성공 모양이어도 닫힌다.
    var bad_status = ok;
    bad_status.status = 6;
    try std.testing.expect(!deriveSmokeStatus(bad_status).glyph_texture_uploaded);

    // upload는 표시됐지만 readback 산출물이 없는 경우.
    var no_readback = ok;
    no_readback.texture_readback = 0;
    try std.testing.expect(!deriveSmokeStatus(no_readback).glyph_texture_uploaded);

    // readback은 됐지만 실패 카운터가 올라간 경우에도 닫혀야 한다(GPU readback 신뢰 불가).
    var readback_failed = ok;
    readback_failed.readback_failures = 1;
    try std.testing.expect(!deriveSmokeStatus(readback_failed).glyph_texture_uploaded);

    // upload가 일어나지 않아 upload_bytes=0이면 닫힌다(GPU에 올린 byte가 없다).
    var no_upload_bytes = ok;
    no_upload_bytes.texture_uploaded = 0;
    no_upload_bytes.upload_bytes = 0;
    try std.testing.expect(!deriveSmokeStatus(no_upload_bytes).glyph_texture_uploaded);

    // CPU raster가 0-ink면 source_rasterized=false이고 gate도 false다.
    var no_ink = ok;
    no_ink.source_non_clear_pixels = 0;
    no_ink.texture_non_clear_pixels = 0;
    try std.testing.expect(!deriveSmokeStatus(no_ink).source_rasterized);
    try std.testing.expect(!deriveSmokeStatus(no_ink).glyph_texture_uploaded);
}
